File: hash.go
// go run hash.go
// Author: liudng@gmail.com
// 2014-9-23

package main

import (
	"fmt"
	"math/rand"
	"sort"
	"strings"
	"time"
)

func main() {
	totalStr := 200 * 10000
	myArr := make([]string, totalStr)
	totalLength := 0
	input := strings.Split("白日依山尽黄河入海流欲穷千里目更上一层楼危楼高百尺可以摘星辰不感高声语恐惊天上人", "")

	//生成随机字符串, 长度为3~16
	rand.Seed(time.Now().Unix())
	for i := 0; i < totalStr; i++ {
		tempStr := ""
		tempLength := rand.Intn(13) + 3
		for j := 0; j < tempLength; j++ {
			tempStr += input[rand.Intn(len(input))]
		}
		totalLength += tempLength
		myArr[i] = tempStr
	}

	//测试开始
	begin := time.Now()
	Dict := make(map[string]string)
	for _, v := range myArr {
		keySli := strings.Split(v, "")
		sort.Strings(keySli)
		key := strings.Join(keySli, "")
		if _, s := Dict[key]; s == false {
			Dict[key] = v
		}
	}

	end := time.Now()

	fmt.Printf("共计%v万条数据，数据总长度%v, 其中%v条不重复数据\n", totalStr/10000, totalLength, len(Dict))
	//fmt.Printf("%v\n%v\n", begin, end)
	fmt.Printf("完成过滤共耗时%v\n", end.Sub(begin))
}

File: hash.js
// node hash.js
// Author: liudng@gmail.com
// 2014-9-25

var totalStr = 1000*10000
var myArr = []
var totalLength = 0
var input = "白日依山尽黄河入海流欲穷千里目更上一层楼危楼高百尺可以摘星辰不感高声语恐惊天上人".split("")
var i = 0

for (i = 0; i < totalStr; i++){
    var tempStr = ""
    var tempLength = parseInt(Math.random()*13+3)
    for (var j = 0; j < tempLength; j++){
        tempStr += input[parseInt(Math.random()*input.length)]
    }
    totalLength += tempLength
    myArr.push(tempStr)
}

//测试开始
var begin = new Date()
Dict = {}
var uniq = 0
for (i = 0; i < myArr.length; i++ ) {
    var charArr = myArr[i].split("")
    charArr.sort()
    var sKey = charArr.join("")
    if (typeof Dict[sKey] == "undefined") {
        Dict[sKey] = myArr[i]
        uniq++
    }
}

var end = new Date()

console.log("共计" + totalStr / 10000 + "万条数据，数据总长度" + totalLength + "，其中" + uniq + "条不重复数据")
console.log("完成过滤共耗时" + (end.getTime() - begin.getTime()) /1000.000 + "s")
